src/DssSpell.t.sol

```diff
@@ -2105,309 +2134,746 @@ contract DssSpellTest is DSTest, DSMath {
+            bytes32(uint256(keccak256(abi.encode(bytes32("ETH-B"), uint256(2)))) + 3),
+            bytes32(uint256(-1))
+        );
+        hevm.store(
+            address(vat),
+            bytes32(uint256(keccak256(abi.encode(bytes32("ETH-C"), uint256(2)))) + 3),
+            bytes32(uint256(-1))
+        );
+        hevm.store(
+            address(vat),
+            bytes32(uint256(keccak256(abi.encode(bytes32("WBTC-A"), uint256(2)))) + 3),
+            bytes32(uint256(-1))
+        );
+
+        {
+            uint256 ETHIlkAmt = 20 * WAD;
+            uint256 WBTCIlkAmt = 1 * WAD;
+
+            giveTokens(ETH, ETHIlkAmt * 3);
+            giveTokens(WBTC, WBTCIlkAmt / 10 ** (18 - 8));
+
+            ETH.approve(address(joinETHA), ETHIlkAmt);
+            joinETHA.join(address(this), ETHIlkAmt);
+            ETH.approve(address(joinETHB), ETHIlkAmt);
+            joinETHB.join(address(this), ETHIlkAmt);
+            ETH.approve(address(joinETHC), ETHIlkAmt);
+            joinETHC.join(address(this), ETHIlkAmt);
+
+            WBTC.approve(address(joinWBTCA), WBTCIlkAmt / 10 ** (18 - 8));
+            joinWBTCA.join(address(this), WBTCIlkAmt / 10 ** (18 - 8));
+
+            (,uint256 rate, uint256 spot,,) = vat.ilks("ETH-A");
+            vat.frob("ETH-A", address(this), address(this), address(this), int256(ETHIlkAmt), int256(mul(ETHIlkAmt, spot) / rate));
+
+            (, rate, spot,,) = vat.ilks("ETH-B");
+            vat.frob("ETH-B", address(this), address(this), address(this), int256(ETHIlkAmt), int256(mul(ETHIlkAmt, spot) / rate));
+
+            (, rate, spot,,) = vat.ilks("ETH-C");
+            vat.frob("ETH-C", address(this), address(this), address(this), int256(ETHIlkAmt), int256(mul(ETHIlkAmt, spot) / rate));
+
+            (, rate, spot,,) = vat.ilks("WBTC-A");
+            vat.frob("WBTC-A", address(this), address(this), address(this), int256(WBTCIlkAmt), int256(mul(WBTCIlkAmt, spot) / rate));
+        }
+
+        hevm.warp(block.timestamp + 1);
+        jug.drip("ETH-A");
+        jug.drip("ETH-B");
+        jug.drip("ETH-C");
+        jug.drip("WBTC-A");
+
+        uint256 auctionIdETHA = clipETHA.kicks() + 1;
+        uint256 auctionIdETHB = clipETHB.kicks() + 1;
+        uint256 auctionIdETHC = clipETHC.kicks() + 1;
+        uint256 auctionIdWBTCA = clipWBTCA.kicks() + 1;
+
+        dog.bark("ETH-A", address(this), address(this));
+        dog.bark("ETH-B", address(this), address(this));
+        dog.bark("ETH-C", address(this), address(this));
+        dog.bark("WBTC-A", address(this), address(this));
+
+        assertEq(clipETHA.kicks(), auctionIdETHA);
+        assertEq(clipETHB.kicks(), auctionIdETHB);
+        assertEq(clipETHC.kicks(), auctionIdETHC);
+        assertEq(clipWBTCA.kicks(), auctionIdWBTCA);
+
+        hevm.store(
+            address(end),
+            keccak256(abi.encode(address(this), uint256(0))),
+            bytes32(uint256(1))
+        );
+        assertEq(end.wards(address(this)), 1);
```

Comment:
This is such a common operation similar to `giveTokens`. Can you replace with this?

```
function giveAuth (address _base, address target) internal {
        AuthLike base = AuthLike(_base);

        // Edge case - ward is already set
        if (base.wards(target) == 1) return;

        for (int i = 0; i < 100; i++) {
            // Scan the storage for the ward storage slot
            bytes32 prevValue = hevm.load(
                address(base),
                keccak256(abi.encode(target, uint256(i)))
            );
            hevm.store(
                address(base),
                keccak256(abi.encode(target, uint256(i))),
                bytes32(uint256(1))
            );
            if (base.wards(target) == 1) {
                // Found it
                return;
            } else {
                // Keep going after restoring the original value
                hevm.store(
                    address(base),
                    keccak256(abi.encode(target, uint256(i))),
                    prevValue
                );
            }
        }

        // We have failed if we reach here
        assertTrue(false);
    }
```

Reply:
Done